File: wehe/src/main/java/mobi/meddle/wehe/fragment/SettingsFragment.java
package mobi.meddle.wehe.fragment;

import android.content.DialogInterface;
import android.os.Bundle;
import android.widget.EditText;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.preference.EditTextPreference;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.Preference.OnPreferenceChangeListener;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.SwitchPreference;

import mobi.meddle.wehe.R;
import mobi.meddle.wehe.constant.Consts;

/**
 * @author Alankrit Joshi, Derek Ng
 * Settings Fragment controls the app settings and uses most of the elements from Default Shared
 * Preferences
 * Settings item in navigation bar (menu.drawer_view.xml)
 * XML layout: xml.preferences.xml
 */
public class SettingsFragment extends PreferenceFragmentCompat {
    public static final String TAG = "SettingsFragment";

    @Override
    public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
        //attach fragment to xml.preferences
        setPreferencesFromResource(R.xml.preferences, rootKey);
        setCurrentPreferences();
        attachListeners();
    }

    /**
     * Set the summary labels to display the current values of the server, area, and ks2p when
     * loading page or when using defaults.
     */
    private void setCurrentPreferences() {
        ListPreference serverPref
                = getPreferenceScreen().findPreference(getString(R.string.pref_server_key));
        assert serverPref != null;
        serverPref.setSummary(String.format(getString(R.string.pref_cur_server), serverPref.getValue()));

        EditTextPreference areaPref
                = getPreferenceScreen().findPreference(getString(R.string.pref_area_key));
        assert areaPref != null;
        areaPref.setSummary(String.format(getString(R.string.pref_cur_percent),
                Integer.parseInt(areaPref.getText())));

        EditTextPreference ks2pPref
                = getPreferenceScreen().findPreference(getString(R.string.pref_ks2p_key));
        assert ks2pPref != null;
        ks2pPref.setSummary(String.format(getString(R.string.pref_cur_percent),
                Integer.parseInt(ks2pPref.getText())));
    }

    private void attachListeners() {
        //determines if input is valid number between 0 and 100, used for area and ks2p
        OnPreferenceChangeListener numberListener = new OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, @NonNull Object newValue) {
                try {
                    int newNumber = Integer.parseInt(newValue.toString(), 10);
                    if (newNumber >= 0 && newNumber <= 100) {
                        preference.setSummary(String.format(getString(R.string.pref_cur_percent),
                                newNumber));
                        return true;
                    }
                } catch (NumberFormatException ignored) {

                }
                Toast.makeText(getContext(), getString(R.string.inval_percent), Toast.LENGTH_LONG).show();
                return false;
            }
        };

        //set custom server
        ListPreference servPref =
                getPreferenceScreen().findPreference(getString(R.string.pref_server_key));
        assert servPref != null;
        OnPreferenceChangeListener changeCurrentText = new OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, @NonNull Object newValue) {
                int index = servPref.findIndexOfValue(newValue.toString());
                if (index != servPref.getEntries().length - 1) { //not custom server
                    preference.setSummary(String.format(getString(R.string.pref_cur_server),
                            newValue.toString()));
                    return true;
                }

                /*custom server*/
                //custom server text box
                EditText customServer = new EditText(getContext());
                customServer.setText(R.string.cust_server);
                customServer.setHint("");

                //get previous server if user cancels typing in new server
                String oldServer = preference.getSharedPreferences().getString(
                        getString(R.string.pref_server_key), Consts.DEFAULT_SERVER);

                //dialogue to popup to let user type in new server
                new AlertDialog.Builder(requireContext())
                        .setTitle(getString(R.string.cust_server))
                        .setView(customServer)
                        .setPositiveButton(getString(android.R.string.ok),
                                new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        //set new server
                                        String customServerText = customServer.getText().toString().toLowerCase();
                                        if (customServerText.matches("[a-z0-9.-]+")) {
                                            servPref.setValue(customServerText);
                                            preference.setSummary(String.format(getString(R.string.pref_cur_server),
                                                    customServerText));
                                        } else {
                                            Toast.makeText(getContext(), getString(R.string.inval_server),
                                                    Toast.LENGTH_LONG).show();
                                        }
                                    }
                                })
                        .setNegativeButton(getString(android.R.string.cancel),
                                new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        //user cancels; set server to the previous server
                                        servPref.setValue(oldServer);
                                        dialog.cancel();
                                    }
                                })
                        .create().show();

                return true;
            }
        };

        //use default settings switch
        OnPreferenceChangeListener switchListener = new OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, @NonNull Object newValue) {
                boolean defaultSettings = Boolean.parseBoolean(newValue.toString());
                if (!defaultSettings) {
                    return true;
                }

                ListPreference serverPref =
                        getPreferenceScreen().findPreference(getString(R.string.pref_server_key));
                EditTextPreference areaPref =
                        getPreferenceScreen().findPreference(getString(R.string.pref_area_key));
                EditTextPreference ks2pPref =
                        getPreferenceScreen().findPreference(getString(R.string.pref_ks2p_key));

                assert serverPref != null;
                serverPref.setValue(Consts.DEFAULT_SERVER);
                assert areaPref != null;
                areaPref.setText(String.valueOf(Consts.A_THRESHOLD));
                assert ks2pPref != null;
                ks2pPref.setText(String.valueOf(Consts.KS2PVAL_THRESHOLD));

                setCurrentPreferences();

                return true;
            }
        };

        //attach the listeners
        EditTextPreference areaPref =
                getPreferenceScreen().findPreference(getString(R.string.pref_area_key));
        assert areaPref != null;
        areaPref.setOnPreferenceChangeListener(numberListener);

        EditTextPreference ks2pPref =
                getPreferenceScreen().findPreference(getString(R.string.pref_ks2p_key));
        assert ks2pPref != null;
        ks2pPref.setOnPreferenceChangeListener(numberListener);

        ListPreference serverPref =
                getPreferenceScreen().findPreference(getString(R.string.pref_server_key));
        assert serverPref != null;
        serverPref.setOnPreferenceChangeListener(changeCurrentText);

        SwitchPreference defaultSwitch =
                getPreferenceScreen().findPreference(getString(R.string.pref_switch_key));
        assert defaultSwitch != null;
        defaultSwitch.setOnPreferenceChangeListener(switchListener);
    }
}

File: wehe/src/main/java/mobi/meddle/wehe/ui/main/fragments/SettingsFragment.kt
package mobi.meddle.wehe.ui.main.fragments

import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.util.TypedValue
import android.view.ContextThemeWrapper
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.preference.EditTextPreference
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import androidx.preference.SwitchPreference
import dagger.hilt.android.AndroidEntryPoint
import mobi.meddle.wehe.R
import mobi.meddle.wehe.constant.Consts
import java.util.Locale

/**
 * @author Alankrit Joshi, Derek Ng
 * Settings Fragment controls the app settings and uses most of the elements from Default Shared
 * Preferences
 * Settings item in navigation bar (menu.drawer_view.xml)
 * XML layout: xml.preferences.xml
 */
@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat() {
    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        // Set the preference file name
        preferenceManager.sharedPreferencesName = getString(R.string.preference_file_key)

        // Load the preferences from XML
        setPreferencesFromResource(R.xml.preferences, rootKey)

        // Initialize the current preferences display
        setCurrentPreferences()
        attachListeners()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = super.onCreateView(inflater, container, savedInstanceState)

        // Ensure the list view background uses theme colors
        listView?.setBackgroundColor(requireContext().getColorFromAttr(android.R.attr.colorBackground))

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Ensure fragment background uses theme colors
        view.setBackgroundColor(requireContext().getColorFromAttr(android.R.attr.colorBackground))
    }

    // Extension function to get color from theme attribute
    private fun Context.getColorFromAttr(attr: Int): Int {
        val typedValue = TypedValue()
        theme.resolveAttribute(attr, typedValue, true)
        return typedValue.data
    }

    // Extension function to check if dark mode is enabled
    private fun Context.isDarkModeEnabled(): Boolean {
        return resources.configuration.uiMode and
                Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
    }

    /**
     * Set the summary labels to display the current values of the server, area, and ks2p when
     * loading page or when using defaults.
     */
    private fun setCurrentPreferences() {
        val serverPref =
            checkNotNull(preferenceScreen.findPreference<ListPreference>(getString(R.string.pref_server_key)))
        serverPref.summary =
            String.format(getString(R.string.pref_cur_server), serverPref.value)

        val areaPref =
            checkNotNull(preferenceScreen.findPreference<EditTextPreference>(getString(R.string.pref_area_key)))
        areaPref.summary = String.format(
            getString(R.string.pref_cur_percent),
            areaPref.text!!.toInt()
        )

        val ks2pPref =
            checkNotNull(preferenceScreen.findPreference<EditTextPreference>(getString(R.string.pref_ks2p_key)))
        ks2pPref.summary = String.format(
            getString(R.string.pref_cur_percent),
            ks2pPref.text!!.toInt()
        )

        // Set dialog theme for EditTextPreference dialogs
        areaPref.setDialogLayoutResource(R.layout.preference_dialog_edittext)
        ks2pPref.setDialogLayoutResource(R.layout.preference_dialog_edittext)
    }

    private fun attachListeners() {
        //determines if input is valid number between 0 and 100, used for area and ks2p
        val numberListener =
            Preference.OnPreferenceChangeListener { preference, newValue ->
                try {
                    val newNumber = newValue.toString().toInt(10)
                    if (newNumber in 0..100) {
                        preference.summary = String.format(
                            getString(R.string.pref_cur_percent),
                            newNumber
                        )
                        return@OnPreferenceChangeListener true
                    }
                } catch (ignored: NumberFormatException) {
                }
                Toast.makeText(context, getString(R.string.inval_percent), Toast.LENGTH_LONG).show()
                false
            }

        //set custom server
        val servPref = checkNotNull(
            preferenceScreen.findPreference<ListPreference>(getString(R.string.pref_server_key))
        )
        val changeCurrentText =
            Preference.OnPreferenceChangeListener { preference, newValue ->
                val index = servPref.findIndexOfValue(newValue.toString())
                if (index != servPref.entries.size - 1) { //not custom server
                    preference.summary = String.format(
                        getString(R.string.pref_cur_server),
                        newValue.toString()
                    )
                    return@OnPreferenceChangeListener true
                }

//                /*custom server*/
//
//                // Create themed context wrapper for dialog
//                val dialogContext = ContextThemeWrapper(requireContext(), R.style.AlertDialogTheme)
//
//                // Create header TextView with theme-aware styling
//                val headerView = TextView(dialogContext)
//                headerView.text = getString(R.string.custom_server_text)
//                headerView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20f)
//                headerView.setPadding(30, 30, 30, 10)
//
//                // Get theme-aware text color
//                headerView.setTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorPrimary))
//
//                // Custom server text box with theme-aware styling
//                val customServer = EditText(dialogContext)
//                customServer.setText("")
//                customServer.setTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorPrimary))
//                customServer.setHintTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorHint))
//                customServer.setPadding(30, 30, 30, 20)
//                customServer.hint = ""
//
                // Get previous server if user cancels typing in new server
                val oldServer = preference.sharedPreferences!!.getString(
                    getString(R.string.pref_server_key), Consts.DEFAULT_SERVER
                )
//
//                // Create a LinearLayout to hold both views with theme-aware background
//                val container = LinearLayout(dialogContext)
//                container.orientation = LinearLayout.VERTICAL
//                container.setBackgroundColor(dialogContext.getColorFromAttr(android.R.attr.colorBackground))
//                container.addView(headerView)
//                container.addView(customServer)

                // Create themed context wrapper for dialog
                val dialogContext = requireContext()

// Create header TextView with theme-aware styling
                val headerView = TextView(dialogContext)
                headerView.text = getString(R.string.custom_server_text)
                headerView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20f)
                headerView.setPadding(30, 30, 30, 10)
                headerView.setTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorPrimary))

// Custom server text box with theme-aware styling
                val customServer = EditText(dialogContext)
                customServer.setText("")
                customServer.setTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorPrimary))
                customServer.setHintTextColor(dialogContext.getColorFromAttr(android.R.attr.textColorHint))
// Also set background tint to ensure it's visible in all themes
                customServer.setPadding(30, 30, 30, 20)
                customServer.hint = getString(R.string.custom_server_hint) // Add a hint string resource if you don't have one

                val dialogView = layoutInflater.inflate(R.layout.dialog_custom_server, null)
                // Create dialog with appropriate styling
                AlertDialog.Builder(dialogContext)
                    .setView(dialogView)
                    .setPositiveButton(
                        getString(android.R.string.ok)
                    ) { _, _ -> //set new server
                        val customServerText =
                            customServer.text.toString().lowercase(Locale.getDefault())
                        if (customServerText.matches("[a-z0-9.-]+".toRegex())) {
                            servPref.value = customServerText
                            preference.summary = String.format(
                                getString(R.string.pref_cur_server),
                                customServerText
                            )
                        } else {
                            Toast.makeText(
                                context, getString(R.string.inval_server),
                                Toast.LENGTH_LONG
                            ).show()
                        }
                    }
                    .setNegativeButton(
                        getString(android.R.string.cancel)
                    ) { dialog, _ -> //user cancels; set server to the previous server
                        servPref.value = oldServer
                        dialog.cancel()
                    }
                    .create().show()
                true
            }

        //use default settings switch
        val switchListener =
            Preference.OnPreferenceChangeListener { _, newValue ->
                val defaultSettings = newValue.toString().toBoolean()
                if (!defaultSettings) {
                    return@OnPreferenceChangeListener true
                }

                val serverPref =
                    preferenceScreen.findPreference<ListPreference>(getString(R.string.pref_server_key))
                val areaPref =
                    preferenceScreen.findPreference<EditTextPreference>(
                        getString(R.string.pref_area_key)
                    )
                val ks2pPref =
                    preferenceScreen.findPreference<EditTextPreference>(
                        getString(R.string.pref_ks2p_key)
                    )

                checkNotNull(serverPref)
                serverPref.value = Consts.DEFAULT_SERVER
                checkNotNull(areaPref)
                areaPref.text = Consts.A_THRESHOLD.toString()
                checkNotNull(ks2pPref)
                ks2pPref.text = Consts.KS2PVAL_THRESHOLD.toString()

                setCurrentPreferences()
                true
            }

        //attach the listeners
        val areaPref = checkNotNull(
            preferenceScreen.findPreference<EditTextPreference>(getString(R.string.pref_area_key))
        )
        areaPref.onPreferenceChangeListener = numberListener

        val ks2pPref = checkNotNull(
            preferenceScreen.findPreference<EditTextPreference>(getString(R.string.pref_ks2p_key))
        )
        ks2pPref.onPreferenceChangeListener = numberListener

        val serverPref = checkNotNull(
            preferenceScreen.findPreference<ListPreference>(getString(R.string.pref_server_key))
        )
        serverPref.onPreferenceChangeListener = changeCurrentText

        val defaultSwitch = checkNotNull(
            preferenceScreen.findPreference<SwitchPreference>(getString(R.string.pref_switch_key))
        )
        defaultSwitch.onPreferenceChangeListener = switchListener
    }
}